Support for Automatic Plugins

cdecoux opened the pull request:
Packer added support for pulling in external plugins via a `required_plugins` block and `packer init`
You can see the reference docs here: https://www.packer.io/plugins#installing-plugins

There's an existing issue out - #100 - to rename this repo to fit the naming convention. I attempted that in my fork, but turns out there's more stuff needed. I followed the [docs for plugin development](https://www.packer.io/docs/plugins/creation#extending-packer) and referenced a [scaffolding repo provided by Hashicorp](https://github.com/hashicorp/packer-plugin-scaffolding) to get a general idea of what needed to change. I made shortcuts to get a working example, so definitely could be improved upon to make properly updated, but hopefully this PR can be a good starting point for that. 

I've got my fork built, released, and get a successful attempt:

Code:
```hcl
packer {
  required_plugins {
    arm = {
      source = "github.com/cdecoux/builder-arm"
      version = "1.0.0"
    }
  }
}
```

Command:
```sh 
packer init .
```

The added workflow is only triggered if a proper tag is pushed. The tag must be a valid [Semantic Version](https://semver.org/) preceded with a `v`.

As of now, I have **not** tested actually using it. Nor have I tested this with the legacy usage. I'm just now learning about Packer ATM. Those should be done before merging this PR.

mkaczanowski commented:
Hey @cdecoux,

thanks a lot for this PR. I'll check it out in a few days!

cdecoux commented on .goreleaser.yml:
```diff
@@ -0,0 +1,80 @@
+    # consumption by Packer.io. To include a separate docs.zip uncomment the following command.
+    #- make ci-release-docs
+    # Check plugin compatibility with required version of the Packer SDK
+    # - make plugin-check
```
The scaffolding repo (https://github.com/hashicorp/packer-plugin-scaffolding) originally had this `make plugin-check` enabled and required a GNUmakefile from the same repo. I tried that myself and had a few errors, so I've bypassed it for now.

dbast commented:
@cdecoux any idea where to put the binary during CI (no download from release page in that case) so that packer finds it without `packer init`. (That should be possible independent from the repo rename topic)

dbast commented:
hm... the logs tell, that is actually detected `[DEBUG] Discovered plugin: arm = /home/runner/work/packer-builder-arm/packer-builder-arm/packer-builder-arm`... but then fails with
```
Error: Failed to initialize build "arm"

error initializing builder 'arm': plugin exited before we could connect
```

dbast commented:
@cdecoux Finished your PR... fixed the lint findings, added versioning so that `make plugin-check` is happy (also now runs via the CI test flow). 

Also did a release in my **renamed** fork of the repothat can be used for testing via the packer plugin init process:
* by adding the plugin to a `$board.pkr.hcl`
  ```
  packer {
     required_plugins {
       arm = {
         version = ">=v1.1.0"
         source  = "github.com/dbast/arm"
       }
    }
  }
  ```

* calling `packer init $board.pkr.hcl` the plugin gets downloaded and can be used

I removed GPG signing for the time being (not required / afaik not checked by packer) and also disabled the non Linux binary builds (they are not useable). 

Important next step before the merge is renaming the repo to the pattern `packer-plugin-*` and aligning the binary in this PR to be named the same.

cdecoux commented:
> @cdecoux Finished your PR... fixed the lint findings, added versioning so that `make plugin-check` is happy (also now runs via the CI test flow).

Awesome work! If the `test` workflow runs on every version release, then it looks like that can be removed from the gorelease. I had just uncommented it until I saw the test workflow, so I'll go ahead and re-remove it.

goldyfruit commented:
Hi guys, thanks for this PR!, I'm glad to see this change!

shanduur commented on .gitignore:
```diff
@@ -21,3 +21,4 @@
 .packer_plugins
 packer_cache
 packer-builder-arm
```
If changing name of the package, should this be kept?

goldyfruit replied:
> If changing name of the package, should this be kept?

`packer-builder-arm` will not be required anymore I guess.

cdecoux commented:
Synced changes and resolved conflicts. Are we able to get this merged after renaming the repo to `packer-plugin-arm`?

(and FYI, renaming the repo should redirect any referenced of `https://github.com/mkaczanowski/packer-builder-arm` to `https://github.com/mkaczanowski/packer-plugin-arm`)

jayofelony commented:
I thought I try your repo but I get the following error when building an image.

``` Pwnagotchi Torch 64bit.arm.opi-pwnagotchi: mounting binfmt_misc with: [mount -t binfmt_misc binfmt_misc /tmp/1445813868/proc/sys/fs/binfmt_misc]
    Pwnagotchi Torch 64bit.arm.opi-pwnagotchi: mounting sysfs with: [mount -t sysfs sysfs /tmp/1445813868/sys]
==> Pwnagotchi Torch 64bit.arm.opi-pwnagotchi: Failed to find binfmt_misc for /usr/bin/qemu-aarch64-static under /proc/sys/fs/binfmt_misc
```

shanduur commented:
@jayofelony which OS are you using?

jayofelony commented:
I'm using WSL Ubuntu under Windows 11. Trying to build an image for Raspios Bullseye 64bit. When I use the main repo it works like it should.
